doc/SEMANTICS.md

```diff
@@ -74,13 +74,29 @@ Synchronization operations (`fsync`, `fdatasync`) are currently no-ops because w
 
 Space allocation (`fallocate`, `posix_fallocate`) are not supported.
 
+#### Deletes
+
+File deletion (`unlink`) is planned but not yet supported (see [#78](https://github.com/awslabs/mountpoint-s3/issues/78)).
+When `unlink` is implemented, the following semantics are proposed.
+
+For files not yet committed to S3, the client will make a best effort to stop the file being uploaded.
+
+- If the file was not yet opened for writing, the file entry will be deleted.
+- If a file is already open and being written to, the multipart upload will be cancelled and any future bytes written to the file handle will be discarded.
```

Comment:
I think this is a bit too in the weeds (it requires you to know the difference between `mknod` and `open`). I'd just say something like:
> For files not yet committed to S3, the client will make a best effort to stop any inflight uploads, and immediately remove the file from its directory. If there are still open file handles to the file, future writes to them will fail.

Reply:
Thanks, that's much nicer for users of higher-level file interfaces.

Reply:
To clarify on behavior, I think we describe two different things - although I prefer your proposed behavior.

>future writes to [file handles] will fail.

In this case, we are saying after unlink that `libc::EIO` will be returned. I am happy with that, and if there's a use case we can update the logic in future to discard bytes with no error.

Reply:
Yeah, we should definitely fail writes that we know won't be persisted.